core/src/stored/backends/unix_tape_device.cc

```diff
@@ -61,6 +73,37 @@ unix_tape_device::unix_tape_device()
   SetCap(CAP_ADJWRITESIZE); /* Adjust write size to min/max */
 }
 
+ssize_t unix_tape_device::d_read(int fd, void* buffer, size_t count)
+{
+  ssize_t ret = ::read(fd, buffer, count);
+  /* If the driver fails to `read()` with `ENOMEM`, then the provided buffer
+   * was too small. By re-reading with a temporary buffer that is enlarged
+   * step-by-step, we can read the block and returns its first `count` bytes.
+   * This allows the calling code to read the block header and resize its buffer
+   * according to the size recorded in that header.
+   */
+  if (ret == -1 && errno == ENOMEM) {
+    for (auto bufsize : bufsizes(count)) {
+      // first go back one block so we can re-read
+      if(!bsr(1)) {
+        errno = EIO;
+        return ret;
+      }
```

Comment:
This should not return `EIO`, otherwise readblockfromdev might try to accidentally read the next block.